src/Mod/Sketcher/Gui/SketcherTransformationExpressionHelper.cpp

```diff
@@ -0,0 +1,143 @@
+    size_t shapeGeometrySize,
+    int numberOfCopies,
+    int secondNumberOfCopies)
+{
+    // apply stored expressions to new constraints, but bail out if we haven't stored
+    // anything
+    if (originalExpressions.empty() || !sketchObject) {
+        return;
+    }
+
+    std::string sketchObj = Gui::Command::getObjectCmd(sketchObject);
+
+    const std::vector<Sketcher::Constraint*>& vals = sketchObject->Constraints.getValues();
+
+    for (size_t i = 0; i < vals.size(); i++) {
+        const auto& cstr = vals[i];
+        if (cstr->isDriving && cstr->isDimensional()) {
+
+            bool expressionApplied = false;
+
+            for (const auto& exprPair : originalExpressions) {
+                if (expressionApplied) {
+                    break;  // found a match, stop searching for this constraint
+                }
+
+                int originalGeoId = exprPair.first;
+
+                // index of the original geometry before operation of the sketcher tool
+                int originalIndex = indexOfGeoId(listOfGeoIds, originalGeoId);
+                if (originalIndex >= 0) {
+                    int firstCurveCreated = sketchObject->getHighestCurveIndex() + 1
+                        - static_cast<int>(shapeGeometrySize);
+                    int size = static_cast<int>(listOfGeoIds.size());
+
+                    // check all copies of this geometry as we assign them the same expression
+                    int numberOfCopiesToMake = numberOfCopies == 0 ? 1 : numberOfCopies;
+
+                    for (int k = 0; k < secondNumberOfCopies && !expressionApplied; k++) {
+                        for (int copy = 1; copy <= numberOfCopiesToMake && !expressionApplied;
+                             copy++) {
+                            int expectedNewGeoId = firstCurveCreated + originalIndex
+                                + size * (copy - 1) + size * numberOfCopiesToMake * k;
+
+                            // if this constraint references our copied geometry, apply the
+                            // expression
+                            if (cstr->First == expectedNewGeoId
+                                || (cstr->Second == expectedNewGeoId
+                                    && cstr->Type != Sketcher::Radius
+                                    && cstr->Type != Sketcher::Diameter
+                                    && cstr->Type != Sketcher::Weight)) {
+
+                                Gui::Command::doCommand(Gui::Command::Doc,
+                                                        "%s.setExpression('Constraints[%d]', '%s')",
+                                                        sketchObj.c_str(),
+                                                        static_cast<int>(i),
+                                                        exprPair.second->toString().c_str());
+                                expressionApplied = true;
+                            }
+                        }
+                    }

```

Comment:
I suggest extracting this to a helper function: the level of nesting here is sort of bonkers. Feel free to find a different seam, of course.